File: Basic-Py/1-Randomise.py
# Week 1 - 1) Write a function that randomly shuffles the array of integers and
#             explain the rationale behind its implementation.

# Sets the imports in which the program uses.
from random import *

def Randomise(array):
    ''' Randomises the elements in a passed list.'''
    # Set random list in which to append.
    randomArray = []
    
    while len(array) != 0:

        # Generate a random number between 0 and length of array.
        randomInt = randint(0, len(array)-1)

        # Appends the value of the index randomInt in array to randomList
        randomArray.append(array[randomInt])

        # Remove the value that was selected from the array.
        array.remove(array[randomInt])

    return randomArray

if __name__ == "__main__":
    while True:
        try:
            # Create an array from the users input based off of a comma.
            array = input("Enter an array (seperate by a ','): ").split(",")

            # Converts all values in the array to integers.
            array = [int(n) for n in array]

            # Outputs the returned value from Randomise.
            print(Randomise(array))
        except:
            break

'''
Explain the Rationale behind its implementation.
------------------------------------------------
This could be implemented in generating a collection of dice rolls, or generate a
random order in which people are placed in a competition with the first index
representing first place.
'''

File: Basic-Cpp/1-Randomise.cpp
/* 
Week 1 - 1) Write a function that randomly shuffles the array of integers and
            explain the rationale behind its implementation.
*/

#include <iostream>
#include <cstdlib>
#include <vector>
using namespace std;

vector<int> Randomise(vector<int> &data) {
    /* Function to randomise the order of a vector. */
    int index = 0; /* Default index to first item */
    int length = data.size(); /* Get length of vector. */
    int randomIndex = 0; /* Set default randomIndex as 0 */
    vector<int> randomList;
    while (index++ < length){
        randomIndex = (rand()%(data.size())); /* Generate a random value of an index in the array. */
        randomList.insert(randomList.begin(),data[randomIndex]); /* Insert in new vector randomList particular vector. */
        data.erase(data.begin()+randomIndex); /* Erase the item that has been inserted so it isn't picked again. */
    }
    
    return randomList;
}

int main() {
    /* Predefine creation of vector. */
    int value, number;
    vector<int> data;
    
    
    cout << "Enter length of list: ";
    cin >> value;
    
    /* Input numbers that are appended to a vector. */
    for (int i = 0; i < value; i++) {
        cout << "Enter a number: ";
        cin >> number;
        data.insert(data.begin()+i,number);
    }
    
    vector<int> randomResult = Randomise(data);
    for (int i = 0; i < randomResult.size(); i++){
        cout << randomResult[i] << " ";
    }
    cout << endl;
    return 0;
}
